File: day2/day2.exs
values = %{"A" => 0, "B" => 1, "C" => 2, "X" => 1, "Y" => 2, "Z" => 3}

File.read!("input.txt")
|> String.split("\n")
|> tap(
  &(Enum.reduce(
      &1,
      0,
      fn <<a::binary-size(1), " ", b::binary>> = ab, acc ->
        acc + values[b] +
          cond do
            ab == "A Y" -> 6
            ab == "B Z" -> 6
            ab == "C X" -> 6
            values[a] + 1 == values[b] -> 3
            true -> 0
          end
      end
    )
    |> IO.inspect(label: "Part 1"))
)
|> tap(
  &(Enum.reduce(
      &1,
      0,
      fn <<a::binary-size(1), " ", b::binary>>, acc ->
        acc +
          case b do
            "X" -> Integer.mod(values[a] - 1, 3) + 1
            "Y" -> values[a] + 1 + 3
            "Z" -> Integer.mod(values[a] + 1, 3) + 1 + 6
          end
      end
    )
    |> IO.inspect(label: "Part 2"))
)

File: day2/day2.py
p1 = {
    "A": {
        "X": 4,
        "Y": 8,
        "Z": 3,
    },
    "B": {
        "X": 1,
        "Y": 5,
        "Z": 9,
    },
    "C": {
        "X": 7,
        "Y": 2,
        "Z": 6,
    },
}
p2 = {
    "A": {
        "X": 3,
        "Y": 4,
        "Z": 8,
    },
    "B": {
        "X": 1,
        "Y": 5,
        "Z": 9,
    },
    "C": {
        "X": 2,
        "Y": 6,
        "Z": 7,
    },
}

with open("input.txt") as f:
    xs = f.read().strip().split("\n")
    print(f"Part 1: {sum([p1[l[0]][l[-1]] for l in xs])}")
    print(f"Part 2: {sum([p2[l[0]][l[-1]] for l in xs])}")
